engine/table/src/main/java/io/deephaven/engine/util/AbstractScriptSession.java

```diff
@@ -96,31 +98,25 @@ public ExecutionContext getExecutionContext() {
     }
 
     protected synchronized void publishInitial() {
-        try (S empty = emptySnapshot(); S snapshot = takeSnapshot()) {
-            applyDiff(empty, snapshot, null);
-        }
-    }
-
-    protected interface Snapshot extends SafeCloseable {
-
+        lastSnapshot = emptySnapshot();
+        observeScopeChanges();
     }
 
     @Override
-    public synchronized SnapshotScope snapshot(@Nullable SnapshotScope previousIfPresent) {
-        // TODO deephaven-core#2453 this should be redone, along with other scope change handling
-        if (previousIfPresent != null) {
-            previousIfPresent.close();
-        }
-        S snapshot = takeSnapshot();
-        return () -> finishSnapshot(snapshot);
+    public void observeScopeChanges() {
+        observeScopeChangesInternal(null);
     }
 
-    private synchronized void finishSnapshot(S beforeSnapshot) {
-        try (beforeSnapshot; S afterSnapshot = takeSnapshot()) {
-            applyDiff(beforeSnapshot, afterSnapshot, null);
+    private synchronized Changes observeScopeChangesInternal(@Nullable final RuntimeException evaluationError) {
+        final S beforeSnapshot = lastSnapshot;
+        lastSnapshot = takeSnapshot();
+        try (beforeSnapshot) {
+            return applyDiff(beforeSnapshot, lastSnapshot, evaluationError);
         }
     }
 
+    protected interface Snapshot extends SafeCloseable {
```

Comment:
Thought (that shouldn't preclude merging): If this is protected and we're removing the generics and other members, do we need generics anymore? As it is protected and not accessible from any non-protected methods, I don't think we need to consider it at all.

Consider: initialSnapshot only runs in the constructor anyway (obligatory "don't call abstract/non-final methods in a constructor), and takeSnapshot()/applyDiff() is only called by observeScopeChangesInternal.. I haven't gone through the exercise yet, but it might be possible to get rid of most of this, pushing it down to just part of the Snapshot interface/implementation, and let the script session not care/cast about what type it has (but the subclass would work out specifics).

Reply:
Let's defer this to a subsequent issue.